Home 06

valed-dm opened the pull request:
Приложение соответствует критериям ДЗ.

valed-dm committed:
README INFO

valed-dm committed:
.gitignore, .dockerignore

valed-dm committed:
requirements.txt

valed-dm committed:
docker-compose.yaml

valed-dm committed:
Dockerfile

valed-dm committed:
config.py

valed-dm committed:
wsgi.py

valed-dm committed:
prestart.sh

valed-dm committed:
main.py

valed-dm committed:
db.py

valed-dm committed:
crude files

valed-dm committed:
db models files

valed-dm committed:
api files

valed-dm committed:
helper functions

valed-dm committed:
book schema

valed-dm committed:
css, bootstrap

valed-dm committed:
images project assets

valed-dm committed:
js to project

valed-dm committed:
macros to project

valed-dm committed:
app html pages added

valed-dm committed:
flask db init

valed-dm committed:
flask db migrate
init gbooks postgres database

valed-dm committed:
adminer credentials

valed-dm committed:
.env.sample

valed-dm committed:
how to start app

mahenzon commented on homework_06/READ_COMMENTS.md:
```diff
@@ -0,0 +1,65 @@
+Сложности в реализации:
+
+- импорты python - зависят от правильной структура проекта - здесь нужен квалифицированный совет
+- pycharm professional в проекте ругается на импорты в /app/database/crud - from database import db, from books.schemas import Book - но все работает отлично
```
надо кликнуть правой клавишей мыши на папку `app` и внизу выбрать Mark directory as -> sources root

valed-dm replied:
спасибо

mahenzon commented on homework_06/READ_COMMENTS.md:
```diff
@@ -0,0 +1,65 @@
+- хотелось, чтобы ORM автоматически заполнял все таблицы, но пришлось сделать это вручную
+- получение данных (many-to-many) - в модели добавил lazy=«selectin», но до конца не разобрался в этом механизме
+- выборка данных из БД /app/database/crud/get_books.py - не понятен механизм joined load-subquery
+- этот запрос к одной (many-to-many) связи Book.authors возвращает также данные другой (many-to-many) связи  Book.categories и (one-to-one) связи - Book.imd_src. Именно это мне и надо, но непонятна сама магия.
```
сразу подгружает? или при обращении подтягивает?

valed-dm replied:
при запросе к БД

mahenzon commented on homework_06/READ_COMMENTS.md:
````diff
@@ -0,0 +1,65 @@
+Вопросы остались, например не получилось использовать insert on conflict:
+<https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert>
+
+```code
````
тут вместо `code` можно написать `python`, тогда в readme будет выделяться синтаксис

valed-dm replied:
спасибо

mahenzon commented on homework_06/READ_COMMENTS.md:
```diff
@@ -0,0 +1,65 @@
+        constraint='pk_my_table',
+        set_=dict(data='updated value')
+    )
+    print(do_update_stmt)
```
выражение составлено, с ним всё ок. остаётся его выполнить. нужно сделать

```python
session.execute(do_update_stmt)
```

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+def add_book(book, rubric, remarks):
+    b = book
+    authors = authors_string(b.authors)
+    library_idx = db.session.query(Book.google_book_id).all()
```
так как любой запрос из базы даных это лишняя нагрузка, то тут будет узкое горлышко. дальше покажу как тут лучше сделать

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+    b = book
+    authors = authors_string(b.authors)
+    library_idx = db.session.query(Book.google_book_id).all()
+    library_idx = [item[0] for item in library_idx]
```
чтобы тут не перекладывать элементы из тюплов можно воспользоваться методом `scalars`

```python
from sqlalchemy import select

all_books_ids = db.session.scalars(select(Book.id)).all()
```

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+    library_idx = [item[0] for item in library_idx]
+
+    # prepares book data to be filled in books table
+    if b.google_book_id not in library_idx:
```
вот так можно проверить на наличие, будет отправлен один короткий запрос в БД:

```python
from sqlalchemy import exists

id_exists = db.session.query(exists().where(Book.id == google_book_id)).scalar()
# True / False
```

Либо через count сделать можно

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+
+        # fills in the authors table
+        db_authors = db.session.query(Author).all()
+        db_authors = [item.name.lower() for item in db_authors]
```
имена можно вытащить по той же схеме что и айдишники. и база данных может сразу привести к нижнему регистру:

```python
from sqlalchemy import func, select

names_lower = db.session.scalars(select(func.lower(Product.name))).all()
# ['john', 'sam']
```

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+        db_categories = db.session.query(Category).all()
+        db_categories = [
+            category for category in db_categories if category.name in b.categories
+        ]
```
всю фильтрацию лучше делать на уровне базы данных. вдруг добавить надо только одну категорию, а из базы будет вытащено сто

mahenzon commented on homework_06/app/database/crud/add_book.py:
```diff
@@ -0,0 +1,75 @@
+                db.session.add(category_for_book)
+
+        # fills in the book-author association table
+        db_authors = db.session.query(Author).all()
```
авторы уже запрошены ранее. лучше не делать лишних запросов, можно обратиться к переменной, которая хранит всех вытащенных авторов

mahenzon commented on homework_06/app/gbooks/fetch_data_prepare.py:
```diff
@@ -0,0 +1,43 @@
+           and item['volumeInfo']['publishedDate'] \
+        else '1900-01-01'
+
+    description = item['volumeInfo']['description'] \
```
тут (и выше) на несколько строчек разбит однострочник. получается, от однострочника и смысла не осталось. лучше написать дефолтное значение (например `description = 'no description'`), а затем в блоке `if` проверить условие и обновить значение `description`.

mahenzon commented on homework_06/app/gbooks/get_args_book.py:
```diff
@@ -0,0 +1,7 @@
+    category = req.args.get("category")
+    sort = req.args.get("sort")
+
+    return [title, pace, category, sort]
```
тут лучше вернуть данные обернутые в dataclass. а то сложно запомнить на какой позиции что (и ниже то же самое)

mahenzon commented on homework_06/app/gbooks/helpers/str_to_class.py:
```diff
@@ -0,0 +1,8 @@
+
+def str_to_class(s: str) -> Book:
+    if isinstance(s, str):
+        book = eval(s)
```
никогда в коде не надо использовать `eval`, так как он просто выполняет код в строке. а там может быть что угодно. есть различные способы сериализации / десериализации, например json. можно из него наполнять объекты

mahenzon commented on homework_06/app/main.py:
```diff
@@ -0,0 +1,101 @@
+    return render_template("library_books.html", authors=authors, books=library_books)
+
+
+@app.get("/library/book/", endpoint="library_book")
```
всё что начинается на `/library/` можно вынести в отдельный Blueprint, чтобы не нагромождать main

valed-dm committed:
code -> python

valed-dm committed:
scalars method implemented

valed-dm committed:
single id_exists db check on the fly
instead of loading all ids from db then check

valed-dm committed:
doubled author, category name bug fixed
leading and trailing spaces in api data names removed

valed-dm committed:
title() added to authors_str

valed-dm committed:
select, filter logic moved to db level

valed-dm committed:
do_nothing_stmt applied
to insert in author, category tables

valed-dm committed:
fetch_data_prepare.py refactored
for Book dataclass usage, fetch_result.py altered

valed-dm committed:
GET method req data organized
with dataclass usage

valed-dm committed:
GET method req data
extra files changed

valed-dm committed:
main page button name altered
search title altered

valed-dm committed:
main page button one more time renamed

valed-dm committed:
unused str_to_class.py deleted
forbidden eval function was used inside str_to_class.py

valed-dm committed:
blueprints api, library created
and corresponding urls refactored with api.,lib. prefixes

valed-dm committed:
app menu button's names shortened
API page - API, Personal library - Library

valed-dm committed:
datetime bug fixed
incoming date string cut to len=10 to fix date formatting bug is input date string differ from 2020-01-01 '%Y-%m-%d' format

valed-dm committed:
incoming date string cut to len=10
to fix date formatting bag is input date string differ from 2020-01-01 '%Y-%m-%d' format

valed-dm committed:
Merge remote-tracking branch 'origin/home_06' into home_06

valed-dm committed:
library sort by category adjusted
to db categories live content instead of hardcoded list of cats

valed-dm committed:
schemas.book dataclass corrected
authors, categories types set to tuple[str] instead of previously assigned str.

valed-dm committed:
typo in label for="sort"
in my_books_search.html corrected

valed-dm committed:
app_flashed_info class "flashes"
typo corrected

valed-dm committed:
book_modal_add.html code layout corrected
